src/Microsoft.AspNetCore.WebUtilities/FormReader.cs

```diff
@@ -197,17 +221,185 @@ private void Buffer()
         /// <returns>The collection containing the parsed HTTP form body.</returns>
         public static async Task<Dictionary<string, StringValues>> ReadFormAsync(Stream stream, Encoding encoding, CancellationToken cancellationToken = new CancellationToken())
         {
-            var reader = new FormReader(stream, encoding);
+            using (var reader = new FormReader(stream, encoding))
+            {
+                var accumulator = new KeyValueAccumulator();
+                var pair = await reader.ReadNextPairAsync(cancellationToken);
+                while (pair.HasValue)
+                {
+                    accumulator.Append(pair.Value.Key, pair.Value.Value);
+                    pair = await reader.ReadNextPairAsync(cancellationToken);
+                }
+
+                return accumulator.GetResults();
+            }
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+                _charPool.Return(_buffer);
+            }
+        }
+
+        // Algorithm from https://github.com/dotnet/corefx/blob/ac67ffac987d0c27236c4a6cf1255c2bcbc7fe7d/src/System.Runtime.Extensions/src/System/Net/WebUtility.cs#L378
+        // but accepting StringBuilder
+        private string UrlDecode(StringBuilder value, ArrayPool<byte> bytePool, ArrayPool<char> charPool)
+        {
+            if (value == null)
+            {
+                return null;
+            }
+
+            var count = value.Length;
+            UrlDecoder helper = new UrlDecoder(count, Encoding.UTF8, bytePool, charPool);
+
+            // go through the string's chars collapsing %XX and
+            // appending each char as char, with exception of %XX constructs
+            // that are appended as bytes
+
+            for (var pos = 0; pos < count; pos++)
+            {
+                var ch = value[pos];
+
+                if (ch == '+')
+                {
+                    ch = ' ';
+                }
+                else if (ch == '%' && pos < count - 2)
+                {
+                    var h1 = HexToInt(value[pos + 1]);
+                    var h2 = HexToInt(value[pos + 2]);
+
+                    if (h1 >= 0 && h2 >= 0)
+                    {     // valid 2 hex chars
+                        var b = (byte)((h1 << 4) | h2);
+                        pos += 2;
+
+                        // don't add as char
+                        helper.AddByte(b);
+                        continue;
+                    }
+                }
+
+                if ((ch & 0xFF80) == 0)
+                {
+                    helper.AddByte((byte)ch); // 7 bit have to go as bytes because of Unicode
```

Comment:
Brain explodes... I liked it better when I didn't have to know how the decoder worked.